ledger/src/blockstore.rs

```diff
@@ -473,6 +479,85 @@ impl Blockstore {
     }
 
+    #[cfg(test)]
+    fn put_erasure_meta(
+        &self,
+        erasure_set: ErasureSetId,
+        erasure_meta: &ErasureMeta,
+    ) -> Result<()> {
+        let (slot, fec_set_index) = erasure_set.store_key();
+        self.erasure_meta_cf.put_bytes(
+            (slot, u64::from(fec_set_index)),
+            &bincode::serialize(erasure_meta).unwrap(),
+        )
+    }
+
+    /// Attempts to find the previous consecutive erasure set for `erasure_set`.
+    ///
+    /// Checks the map `erasure_metas`, if not present scans blockstore. Returns None
+    /// if the previous consecutive erasure set is not present in either.
+    fn previous_erasure_set(
+        &self,
+        erasure_set: ErasureSetId,
+        erasure_metas: &mut BTreeMap<ErasureSetId, WorkingEntry<ErasureMeta>>,
+    ) -> Result<Option<ErasureSetId>> {
+        let (slot, fec_set_index) = erasure_set.store_key();
+
+        // Check the previous entry from the in memory map to see if it is the consecutive
+        // set to `erasure set`
+        let candidate_erasure_entry = erasure_metas
+            .range((
+                Bound::Included(ErasureSetId::new(slot, 0)),
+                Bound::Excluded(erasure_set),
+            ))
+            .next_back();
+        let candidate_erasure_set = candidate_erasure_entry
+            .filter(|(_, candidate_erasure_meta)| {
+                candidate_erasure_meta.as_ref().next_fec_set_index() == Some(fec_set_index)
+            })
+            .map(|(candidate_erasure_set, _)| *candidate_erasure_set);
+        if candidate_erasure_set.is_some() {
+            return Ok(candidate_erasure_set);
+        }
+
+        // Consecutive set was not found in memory, scan blockstore for a potential candidate
+        let Some(((_, candidate_fec_set_index), candidate_erasure_meta)) = self
+            .erasure_meta_cf
+            .iter(IteratorMode::From(
+                (slot, u64::from(fec_set_index)),
+                IteratorDirection::Reverse,
+            ))?
+            // `find` here, to skip the first element in case the erasure meta for fec_set_index is already present
+            .find(|((_, candidate_fec_set_index), _)| {
+                *candidate_fec_set_index != u64::from(fec_set_index)
+            })
+            // Do not consider sets from the previous slot
+            .filter(|((candidate_slot, _), _)| *candidate_slot == slot)
+        else {
+            // No potential candidates
+            return Ok(None);
+        };
+        let candidate_fec_set_index = u32::try_from(candidate_fec_set_index)
+            .expect("fec_set_index from a previously inserted shred should fit in u32");
+        let candidate_erasure_set = ErasureSetId::new(slot, candidate_fec_set_index);
+        let candidate_erasure_meta: ErasureMeta = deserialize(candidate_erasure_meta.as_ref())?;
+
+        // Add this candidate to erasure metas to avoid blockstore lookup in future
```

Comment:
This is the only diff from #102 addressing this comment: https://github.com/anza-xyz/agave/pull/102#discussion_r1559826225